tests/kernel/mem_protect/mpu/src/kobject.c

```diff
@@ -0,0 +1,170 @@
+struct k_sem *random_sem_type=NULL;
+
+
+/****************************************************************************/
+void kobject_user_tc1(void *p1, void *p2, void *p3)
+{
+	valid_fault= true;
+	k_sem_take(random_sem_type, K_FOREVER);
+
+
+}
+
+/* Test access to a invalid semaphore who's address is NULL. */
+void kobject_test_1(void *p1, void *p2, void *p3)
+{
+
+	/* _k_object_init(random_sem_type); */
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem,
+			      &kobject_mutex,
+			      random_sem_type, NULL);
+
+	k_thread_user_mode_enter(kobject_user_tc1, NULL, NULL, NULL);
+
+}
+/****************************************************************************/
+void kobject_user_tc2(void *p1, void *p2, void *p3)
+{
+	valid_fault= false;
+	k_sem_give(&kobject_sem);
+
+	/* should cause a fault */
+	valid_fault= true;
+	/* typecasting to override compiler warning */
+	k_sem_take((struct k_sem *)&kobject_mutex, K_FOREVER);
+
+
+}
+
+/* Test if a syscall can take a different type of kobject. */
+void kobject_test_2(void *p1, void *p2, void *p3)
+{
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem,
+			      &kobject_mutex, NULL);
+
+	k_thread_user_mode_enter(kobject_user_tc2, NULL, NULL, NULL);
+
+}
+
+/****************************************************************************/
+void kobject_user_tc3(void *p1, void *p2, void *p3)
+{
+	/* should cause a fault */
+	valid_fault= true;
+	k_sem_give(&kobject_sem);
+}
+
+/* test if a user thread can access a k object that has not be granted to it.*/
+void kobject_test_3(void *p1, void *p2, void *p3)
+{
+	k_thread_access_grant(k_current_get(),
+			      &kobject_mutex, NULL);
+
+	k_thread_user_mode_enter(kobject_user_tc3, NULL, NULL, NULL);
+
+}
+
+/****************************************************************************/
+void kobject_user_test4(void *p1, void *p2, void *p3)
+{
+	/* should cause a fault */
+	if((u32_t)p1 == 1){
+		valid_fault= false;
+	}
+	else {
+		valid_fault= true;
+	}
+	k_sem_give(&kobject_sem);
+}
+
+/* Test access revoke. */
+void kobject_test_4(void *p1, void *p2, void *p3)
+{
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem, NULL);
+
+	k_thread_create(&kobject_test_4_tid,
+			kobject_test_4_stack,
+			KOBJECT_STACK_SIZE,
+			kobject_user_test4,
+			(void *)1, NULL, NULL,
+			0, K_INHERIT_PERMS|K_USER, K_NO_WAIT);
+
+	k_sleep(1);
+	k_object_access_revoke(&kobject_sem, k_current_get());
+
+	k_thread_create(&kobject_test_4_tid,
+			kobject_test_4_stack,
+			KOBJECT_STACK_SIZE,
+			kobject_user_test4,
+			(void *)2, NULL, NULL,
+			0, K_INHERIT_PERMS|K_USER, K_NO_WAIT);
+
+	k_thread_abort(k_current_get());
+
+}
+
+/****************************************************************************/
+/* grant access to all user threads that follow */
+void kobject_user_1_test5(void *p1, void *p2, void *p3)
+{
+	valid_fault = false;
+	k_sem_give(&kobject_sem);
+	k_object_access_grant(&kobject_sem, &kobject_test_5_2_tid);
```

Comment:
Hmm..are we supposed to be able to grant access to a thread to which we have no permissions?

Reply:
no we aren't. i'm confused here.

Reply:
oh this was for the linker issue. I was testing the issue with this code.